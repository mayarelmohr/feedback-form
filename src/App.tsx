import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import "./App.css";
import FeedbackForm from "pages/Home";
import Results from "pages/Results";
import Header from "components/Header";
import { FeedbackListProvider } from "store/context/feedbackProvider";

function App() {
  return (
    <Router>
      <Header />
      <main>
        <FeedbackListProvider>
          <Routes>
            <Route path="/" element={<FeedbackForm />} />
            <Route path="/results" element={<Results />} />
          </Routes>
        </FeedbackListProvider>
      </main>
    </Router>
  );
}

export default App;
