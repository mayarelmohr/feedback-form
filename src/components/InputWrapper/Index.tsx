import { ReactNode, FC } from "react";
import classNames from "classnames/bind";
import styles from "./style.module.css";

const cx = classNames.bind(styles);

export interface InputWrapperProps {
  label: string;
  error?: string;
}

export const InputWrapper: FC<InputWrapperProps & { children: ReactNode }> = ({
  label,
  error,
  children,
}) => {
  return (
    <div className={cx("wrapper")}>
      <label>
        <span className={cx("label")}>
          {label} {error ? <span className={cx("required")}>*</span> : ""}
        </span>
        {children}
      </label>
      {error ? <p className={cx("error-message")}>{error}</p> : null}
    </div>
  );
};
