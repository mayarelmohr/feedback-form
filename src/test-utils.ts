import { screen } from "@testing-library/react";

export const getTextField = (name: string) => {
  return screen.getByRole("textbox", { name: new RegExp(name, "i") });
};
