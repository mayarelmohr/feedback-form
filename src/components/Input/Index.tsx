import { InputHTMLAttributes, FC } from "react";
import { InputWrapper, InputWrapperProps } from "components/InputWrapper";
import classNames from "classnames/bind";
import styles from "./style.module.css";

const cx = classNames.bind(styles);

type InputProps = InputWrapperProps & InputHTMLAttributes<HTMLInputElement>;

const Input: FC<InputProps> = ({ label, type = "text", error, ...rest }) => {
  return (
    <InputWrapper label={label} error={error}>
      <input className={cx("input", error ? "error-input" : "")} {...rest} />
    </InputWrapper>
  );
};

export default Input;
