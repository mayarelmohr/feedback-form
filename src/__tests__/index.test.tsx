import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Index from "pages/Home";
import { getTextField } from "test-utils";
import userEvent from "@testing-library/user-event";
import App from "App";

describe("Home page renders all elements", () => {
  it("Home page matches snapshot", () => {
    const { container } = render(<Index />, { wrapper: BrowserRouter });
    expect(container).toMatchSnapshot();
  });

  it("Home page has heading", () => {
    render(<Index />, { wrapper: BrowserRouter });
    const heading = screen.getByRole("heading", {
      name: /add your feedback/i,
    });
    expect(heading).toBeInTheDocument();
  });
});

describe("Feedback Form", () => {
  it("Form has all inputs rendered", () => {
    render(<Index />, { wrapper: BrowserRouter });
    expect(screen.getByLabelText(/name/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/rating/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/comment/i)).toBeInTheDocument();
    const button = screen.getByText(/submit feedback/i);
    expect(button).toBeInTheDocument();
  });

  it("goes to next page when all fields pass validation and submit feedback", async () => {
    render(<App />);
    userEvent.type(getTextField("name"), "John");
    userEvent.type(getTextField("email"), "john.doe@gmail.com");
    userEvent.type(getTextField("comment"), "Amazing product");
    userEvent.click(screen.getByTestId("2-stars"));
    userEvent.click(screen.getByText(/submit feedback/i));
    const heading = await screen.findByRole("heading", {
      name: /feedback results/i,
    });
    expect(heading).toBeInTheDocument();
  });

  it("has all fields required before submission", () => {
    render(<Index />, { wrapper: BrowserRouter });
    userEvent.click(screen.getByText(/submit feedback/i));
    expect(screen.getByText(/please add your name./i)).toBeInTheDocument();
    expect(screen.getByText(/please add your email./i)).toBeInTheDocument();
    expect(screen.getByText(/please add your comment./i)).toBeInTheDocument();
    expect(screen.getByText(/please add your rating./i)).toBeInTheDocument();
  });

  it("shows error when email is invalid", async () => {
    render(<Index />, { wrapper: BrowserRouter });
    const email = getTextField("email");
    userEvent.type(getTextField("email"), "john.doe");
    fireEvent.focusOut(email);
    const emailError = await screen.getByText(/Email is not valid./i);
    expect(emailError).toBeInTheDocument();
  });
});
