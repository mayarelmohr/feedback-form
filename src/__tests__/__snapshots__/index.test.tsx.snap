// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page renders all elements Home page matches snapshot 1`] = `
<div>
  <form
    class="layout"
  >
    <h2
      class="title"
    >
      Add your feedback here
    </h2>
    <div
      class="wrapper"
    >
      <div
        class="flex"
      >
        <div
          class="wrapper"
        >
          <label>
            <span
              class="label"
            >
              Name
               
            </span>
            <input
              class="input"
              name="name"
              placeholder="Add your name here"
              value=""
            />
          </label>
        </div>
        <div
          class="wrapper"
        >
          <label>
            <span
              class="label"
            >
              Email
               
            </span>
            <input
              class="input"
              name="email"
              placeholder="Add your email here"
              value=""
            />
          </label>
        </div>
        <div
          class="wrapper"
        >
          <label>
            <span
              class="label"
            >
              Rating
               
            </span>
            <fieldset
              class="wrapper"
            >
              <label
                class="radio"
              >
                <input
                  class="input"
                  name="rating"
                  type="radio"
                  value="1"
                />
                <span
                  class="sr-only"
                >
                  1
                   stars
                </span>
                <span
                  aria-hidden="true"
                  data-testid="1-stars"
                  role="img"
                >
                  <svg
                    class="rating"
                  >
                    star.svg
                  </svg>
                </span>
              </label>
              <label
                class="radio"
              >
                <input
                  class="input"
                  name="rating"
                  type="radio"
                  value="2"
                />
                <span
                  class="sr-only"
                >
                  2
                   stars
                </span>
                <span
                  aria-hidden="true"
                  data-testid="2-stars"
                  role="img"
                >
                  <svg
                    class="rating"
                  >
                    star.svg
                  </svg>
                </span>
              </label>
              <label
                class="radio"
              >
                <input
                  class="input"
                  name="rating"
                  type="radio"
                  value="3"
                />
                <span
                  class="sr-only"
                >
                  3
                   stars
                </span>
                <span
                  aria-hidden="true"
                  data-testid="3-stars"
                  role="img"
                >
                  <svg
                    class="rating"
                  >
                    star.svg
                  </svg>
                </span>
              </label>
              <label
                class="radio"
              >
                <input
                  class="input"
                  name="rating"
                  type="radio"
                  value="4"
                />
                <span
                  class="sr-only"
                >
                  4
                   stars
                </span>
                <span
                  aria-hidden="true"
                  data-testid="4-stars"
                  role="img"
                >
                  <svg
                    class="rating"
                  >
                    star.svg
                  </svg>
                </span>
              </label>
              <label
                class="radio"
              >
                <input
                  class="input"
                  name="rating"
                  type="radio"
                  value="5"
                />
                <span
                  class="sr-only"
                >
                  5
                   stars
                </span>
                <span
                  aria-hidden="true"
                  data-testid="5-stars"
                  role="img"
                >
                  <svg
                    class="rating"
                  >
                    star.svg
                  </svg>
                </span>
              </label>
            </fieldset>
          </label>
        </div>
      </div>
      <div
        class="wrapper"
      >
        <label>
          <span
            class="label"
          >
            Comment
             
          </span>
          <textarea
            class="textarea"
            name="comment"
            placeholder="Add your comment here"
          />
        </label>
      </div>
    </div>
    <button
      class="button"
      type="submit"
    >
      Submit Feedback
    </button>
  </form>
</div>
`;
